File: src/_03_BehavioralPattern/_03_07_Observer/ChatServer.js
/**
 * @typedef { import("./Subscriber").default } Subscriber
 */

export default class ChatServer {
  /** @type {{ [key: string]: Subscriber[] }} */
  #subscribers;

  constructor() {
    this.#subscribers = {};
  }

  /**
   * @param { string } subject 
   * @param { Subscriber } subscriber 
   */
  subscribe(subject, subscriber) {
    this.#subscribers[subject]
      ? this.#subscribers[subject].push(subscriber)
      : this.#subscribers[subject] = [subscriber]
  }

  /**
   * @param { string } subject 
   * @param { Subscriber } subscriber 
   */
  unsubscribe(subject, subscriber) {
    if (!this.#subscribers[subject]) return

    const idx = this.#subscribers[subject].findIndex(subscriber);
    !isNil(idx) && this.#subscribers[subject].splice(idx, 1);
  }

  /**
   * 
   * @param { string } subject 
   * @param { Subscriber } subscriber 
   * @param { string } message 
   */
  sendMessage(subject, subscriber, message) {
    if (!this.#subscribers[subject]) return

    const resultMessage = `[작성자: ${subscriber.getName()}] - ${message}`;
    this.#subscribers[subject].forEach(s => s.handleMessage(resultMessage));
  }
}

File: src/_03_BehavioralPattern/_03_07_Observer/java/after/ChatServer.java
package _03_BehavioralPattern._03_07_Observer.java.after;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ChatServer {
  
  private Map<String, List<Subscriber>> subscribers = new HashMap<>();

  public void subscribe(String subject, Subscriber subscriber) {
    if (this.subscribers.containsKey(subject)) {
      this.subscribers.get(subject).add(subscriber);
    } else {
      List<Subscriber> list = new ArrayList<Subscriber>();
      list.add(subscriber);
      this.subscribers.put(subject, list);
    }
  }

  public void unsubscribe(String subject, Subscriber subscriber) {
    if (this.subscribers.containsKey(subject)) {
      this.subscribers.get(subject).remove(subscriber);
    }
  }

  public void sendMessage(Subscriber subscriber, String subject, String message) {
    if (this.subscribers.containsKey(subject)) {
      String resultMessage = "[작성자: " + subscriber.getName() + "] " + message;
      this.subscribers.get(subject).forEach(s -> s.handleMessage(resultMessage));
    }
  }
  
}
